### PR TITLE
[develop < T0040-GA] Add index on label

### PR DESCRIPTION
**Pull request by katarinasupe**
### Description

With this new feature, it's possible to create an index on the label. Up until now, it was only possible to create an index on certain node properties. 

Documentation: How-to guide on Docs https://github.com/memgraph/docs/pull/339

### Pull request type

- [x] Feature

### Related issues

Closes https://github.com/memgraph/gqlalchemy/issues/102

######################################

### Reviewer checklist (the reviewer checks this part)
- [x] Core feature implementation
- [x] Tests
- [x] Code documentation
- [x] Documentation on [memgraph/docs](https://github.com/memgraph/docs)

######################################

### CODE REVIEW

File: gqlalchemy/models.py [hunk 1]
```diff
@@ -365,12 +367,20 @@ def field_in_superclass(field, constraint):
         else:
             cls.labels = {cls.label}
 
+        db = kwargs.get("db", None)
```
Comment (BorisTasevski) [inline gqlalchemy/models.py:370]: remove `None`

File: gqlalchemy/models.py [hunk 2]
```diff
@@ -365,12 +367,20 @@ def field_in_superclass(field, constraint):
...
+        if cls.index is True:
+            if db is None:
+                raise GQLAlchemyDatabaseMissingInNodeClassError(cls=cls)
```
Comment (BorisTasevski) [inline gqlalchemy/models.py:373]: We'll discuss on newlines after ifs

File: gqlalchemy/models.py [hunk 3]
```diff
@@ -365,12 +367,20 @@ def field_in_superclass(field, constraint):
...
+            index = MemgraphIndex(cls.label)
+            db.create_index(index)
+
```
Comment (BorisTasevski) [inline gqlalchemy/models.py:376]: Because You have newline after if here

File: gqlalchemy/models.py [hunk 4]
```diff
@@ -365,12 +367,20 @@ def field_in_superclass(field, constraint):
...
         for field in cls.__fields__:
             attrs = cls.__fields__[field].field_info.extra
             field_type = cls.__fields__[field].type_.__name__
 
             label = attrs.get("label", cls.label)
-            db = attrs.get("db", None)
+            if db is None:
+                db = attrs.get("db", None)
```
Comment (BorisTasevski) [inline gqlalchemy/models.py:383]: And again, here You don't have it

File: tests/integration/test_index.py [hunk 5]
```diff
@@ -14,6 +14,72 @@
...
+        id: str
+
+    actual_index = memgraph.get_indexes()
+
+    assert set(actual_index) == {MemgraphIndex("User")}
+
+
+def test_missing_db_in_node_class(memgraph):
+    with pytest.raises(GQLAlchemyDatabaseMissingInNodeClassError):
+
```
Comment (BorisTasevski) [inline tests/integration/test_index.py:32]: remove newline

File: tests/integration/test_index.py [hunk 6]
```diff
@@ -14,6 +14,72 @@
...
+        class User(Node, index=True):
+            id: str
+
+
+def test_db_in_node_class(memgraph):
+    class User(Node, db=memgraph):
```
Comment (BorisTasevski) [inline tests/integration/test_index.py:38]: Can we promote these in-test defined classes to fixtures, and use them somehow as parameters for the tests?

File: tests/integration/test_index.py [hunk 7]
```diff
@@ -14,6 +14,72 @@
...
+        id: str = Field(index=True)
+
+    actual_index = memgraph.get_indexes()
+    assert set(actual_index) == {MemgraphIndex("User", "id")}
+
+
+def test_db_in_node_and_property(memgraph):
+    class User(Node, db=memgraph):
```
Comment (BorisTasevski) [inline tests/integration/test_index.py:46]: Can we promote these in-test defined classes to fixtures, and use them somehow as parameters for the tests?

File: tests/integration/test_index.py [hunk 8]
```diff
@@ -14,6 +14,72 @@
...
+        id: str = Field(index=True, db=memgraph)
+
+    actual_index = memgraph.get_indexes()
+    assert set(actual_index) == {MemgraphIndex("User", "id")}
+
+
+def test_index_on_label_and_property(memgraph):
+    class User(Node, index=True, db=memgraph):
```
Comment (BorisTasevski) [inline tests/integration/test_index.py:54]: Can we promote these in-test defined classes to fixtures, and use them somehow as parameters for the tests?

File: tests/integration/test_index.py [hunk 9]
```diff
@@ -14,6 +14,72 @@
...
+        id: str = Field(index=True, db=memgraph)
+
+    actual_index = memgraph.get_indexes()
+    assert set(actual_index) == {MemgraphIndex("User", "id"), MemgraphIndex("User")}
+
+
+def test_false_index_in_node_class(memgraph):
+    class User(Node, index=False, db=memgraph):
```
Comment (BorisTasevski) [inline tests/integration/test_index.py:62]: Can we promote these in-test defined classes to fixtures, and use them somehow as parameters for the tests?

File: tests/integration/test_index.py [hunk 10]
```diff
@@ -14,6 +14,72 @@
...
+        id: str
+
+    actual_index = memgraph.get_indexes()
+    assert set(actual_index) == set()
+
+
+def test_false_index_no_db_in_node_class(memgraph):
+    class User(Node, index=False):
```
Comment (BorisTasevski) [inline tests/integration/test_index.py:70]: Can we promote these in-test defined classes to fixtures, and use them somehow as parameters for the tests?

File: tests/integration/test_index.py [hunk 11]
```diff
@@ -14,6 +14,72 @@
...
+        id: str
+
+    actual_index = memgraph.get_indexes()
+    assert set(actual_index) == set()
+
+
+def test_false_index_with_db_in_node_class(memgraph):
+    class User(Node, index=False, db=memgraph):
```
Comment (BorisTasevski) [inline tests/integration/test_index.py:78]: Can we promote these in-test defined classes to fixtures, and use them somehow as parameters for the tests?

**Review by BorisTasevski — CHANGES_REQUESTED** [review]

File: tests/integration/test_index.py (hunk 7 above)
Comment (katarinasupe) [reply tests/integration/test_index.py:46]: We couldn't before, the fixture wasn't working alright. @g-despot did we fix this?

File: tests/integration/test_index.py (hunk 5 above)
Comment (katarinasupe) [reply tests/integration/test_index.py:32]: Black did this one and I think it has to stay to pass the tests

File: tests/integration/test_index.py (hunk 7 above)
Comment (g-despot) [reply tests/integration/test_index.py:46]: I think it's possible but I don't see much benefit from it.

File: tests/integration/test_index.py (hunk 5 above)
Comment (BorisTasevski) [reply tests/integration/test_index.py:32]: You are right, my bad

File: tests/integration/test_index.py (hunk 7 above)
Comment (BorisTasevski) [reply tests/integration/test_index.py:46]: There is a huge amount of code duplication and maintenance would be a monstrous nuisance if we leave it as is.

File: gqlalchemy/models.py (hunk 4 above)
Comment (Josipmrden) [reply gqlalchemy/models.py:383]: I'm missing something, where should the newline be?

**Review by g-despot — APPROVED** [review]

File: gqlalchemy/models.py [hunk 12]
```diff
@@ -365,20 +368,32 @@ def field_in_superclass(field, constraint):
...
         for field in cls.__fields__:
             attrs = cls.__fields__[field].field_info.extra
             field_type = cls.__fields__[field].type_.__name__
-
             label = attrs.get("label", cls.label)
-            db = attrs.get("db", None)
             skip_constraints = False
+
+            if db is None:
+                db = attrs.get("db", None)
```
[the remark this answers is not in the document]
Comment (katarinasupe) [reply gqlalchemy/models.py:387]: Great, thanks, will fix this

**Review by BorisTasevski — APPROVED** [review]

**Pull request merged**